traced/traced.go

```diff
@@ -154,24 +188,64 @@ func (tc *TraceCollector) writeFile(out io.WriteCloser) (result error) {
 		// with our local buffer (trimmed to 0-length). On the next loop
 		// iteration, the same will occur, so we're effectively swapping buffers
 		// continuously.
+		//
+		// Since we expect the sources slice to be small, we don't need the same
+		// dance. We just make the sources in state slice nil.
 		tc.mx.Lock()
+		// buffer dance.
 		tmp := tc.buf
 		tc.buf = buf[:0]
 		buf = tmp
+		// reset sources.
+		sources := tc.sources
+		tc.sources = nil
 		tc.mx.Unlock()
 
+		var (
+			currSourceIdx int
+			currSource    = &sources[0]
+		)
 		for i, evt := range buf {
+			if i > currSource.end {
+				// finished the range for this source, move to next.
+				// no need to do any slice boundary check where because
+				// consistency is guaranteed.
+				currSourceIdx++
+				currSource = &sources[currSourceIdx]
+			}
+
+			buf[i] = nil
+
+			// write to the binary output.
 			err := w.WriteMsg(evt)
 			if err != nil {
 				logger.Errorf("error writing event trace: %s", err)
 				return err
 			}
 
-			buf[i] = nil
+			if jsonEncoder == nil {
+				continue
+			}
+
+			// convert the event to a map[string]interface{} for easier manipulation.
+			m := structMap(evt)
+
+			// transform the event.
+			m = transformRec(m)
+
+			// enrich with the IP address from the multiaddr.
+			if ipaddr, err := manet.ToIP(currSource.maddr); err == nil {
+				m["ip"] = ipaddr.String()
+			}
+
+			if err = jsonEncoder.Encode(m); err != nil {
+				logger.Errorf("error writing JSON trace: %s", err)
+				return err
```

Comment:
this return may be problematic, as it breaks the invariants for traces; maybe we should just swallow and log?